Issue 461

hassanelsheikha opened the pull request:
## Link to Github issue or other discussion

This PR stems from #461 but is also pertinent to #591 and #76

## What does this PR do?

Please kindly see documentation [here.](https://github.com/hassanelsheikha/islandora_workbench_docs/blob/1f7b4edd7e1e5435092cd989cbfd75dc42880a63/docs/update_media.md)

## What changes were made?

The `update_media` placeholder function was implemented and related code was updated.

## How to test / verify this PR?

Integration/functional tests are required. 
Manual tests can be conducted by creating a CSV files with the relevant fields and a configuration file that specifies "update_media" as the `task` setting.

Please let me know of any bugs and I will be happy to fix them right away!

## Interested Parties

@mjordan @Natkeeran

## Checklist

* [Y] Before opening this PR, have you opened an issue explaining what you want to to do?
* [Y] Have you run `pycodestyle --show-source --show-pep8 --ignore=E402,W504 --max-line-length=200 yourfile.py`? 
* [Y] Have you included same configuration and/or CSV files useful for testing this PR?
* [N] Have you written unit or integration tests if applicable?
* [N] If the changes in this PR require an additional Python library, have you included it in `setup.py`?
* [N] If the changes in this PR add a new configuration option, have you provided a default for when the option is not present in the .yml file?

hassanelsheikha commented:
I will push out a commit that fixes a bug very soon, as it seems like not putting the line `standalone_media_url: true` causes some issues. For now, could you please test the code with this line in the config?

mjordan commented:
@hassanelsheikha I've started testing and everything works as advertised! I found a few errors and have a few suggestions for bringing the UX closer to that for updating nodes. I don't have enough time right now to report them all thoroughly, but here's one error you can probably fix fairly easily. I'll report the other things this evening.

When performing an operation other than replacing a file, during `--check` I see the following error:

```
./workbench --config update_media.yml --check
OK, connection to Drupal at https://islandora.traefik.me verified.
OK, configuration file has all required values (did not check for optional values).
OK, CSV file input_data/update_media_replace_file.csv found.
OK, all 1 rows in the CSV file have the same number of columns as there are headers (2).
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /tmp/islandora_workbench/./workbench:1861 in <module>                                            │
│                                                                                                  │
│   1858 │   │   │   if config['task'] == 'create_from_files':                                     │
│   1859 │   │   │   │   check_input_for_create_from_files(config, args)                           │
│   1860 │   │   │   else:                                                                         │
│ ❱ 1861 │   │   │   │   check_input(config, args)                                                 │
│   1862 except KeyboardInterrupt:                                                                 │
│   1863 │   print('Exiting before entire --check completed.')                                     │
│   1864 │   logging.warning('Workbench exiting after receiving "ctrl-c" during --check.')         │
│                                                                                                  │
│ /tmp/islandora_workbench/workbench_utils.py:1858 in check_input                                  │
│                                                                                                  │
│   1855 │   │   │   config['id_field'] = 'media_id'                                               │
│   1856 │   │   file_check_csv_data = get_csv_data(config)                                        │
│   1857 │   │   for count, file_check_row in enumerate(file_check_csv_data, start=1):             │
│ ❱ 1858 │   │   │   file_check_row['file'] = file_check_row['file'].strip()                       │
│   1859 │   │   │   # Check for empty 'file' values.                                              │
│   1860 │   │   │   if len(file_check_row['file']) == 0:                                          │
│   1861 │   │   │   │   message = 'CSV row with ID ' + file_check_row[config['id_field']] + ' co  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'file'
```

In "create" tasks, the "file" column is required, so I think it would be OK to also make the "file" column required in "update_media" tasks (which --check suggests) but if that's the case, your docs should say that. But if we require "file", you'll need to incorporate the "allow_missing_files" config setting into your logic, and users will need to set "allow_missing_files" to `true` if they want to perform non-file-replace operations. Otherwise, we'll need separate "update_media" and "replace_file" tasks. Does that make sense?

mjordan commented:
Or alternatively, it might be simpler to not require the `file` column. If it's not present, the task doesn't replace any files. Maybe that's the better route to take.

hassanelsheikha commented:
Thanks for pointing this out. The original goal was to follow the second approach, but it seems like I forgot to add an additional if statement in the code that checks if the file exists. Let me see what I can do.

mjordan commented:
Very good. Second approach provides a much simpler UX. Thanks!

hassanelsheikha commented:
No worries. I (hopefully) fixed the issue but I'm running into some issues with the style checker. I want to place the following `if` statement in the code

```python
if (config['task'] == 'create' or config['task'] == 'add_media' or config['task'] == 'update_media' and 'file' in csv_column_headers) and config['nodes_only'] is False and config['paged_content_from_directories'] is False:
```

Obviously, this is way too long of a line. But I'm not sure how to break it up into separate lines without getting an error along the lines of:

```
    Continuation lines indentation.

    Continuation lines should align wrapped elements either vertically
    using Python's implicit line joining inside parentheses, brackets
    and braces, or using a hanging indent.

    When using a hanging indent these considerations should be applied:
    - there should be no arguments on the first line, and
    - further indentation should be used to clearly distinguish itself
      as a continuation line.
```
Do you have any suggestions for how to fix this?

mjordan commented:
You could nest the ifs like this:

```
if (config['task'] == 'create' or config['task'] == 'add_media' or config['task'] == 'update_media':
    if 'file' in csv_column_headers) and config['nodes_only'] is False and config['paged_content_from_directories'] is False:
```

hassanelsheikha commented:
Ok great, thank you! I was hesitant on doing that at first since it adds a whole new indentation level, but it doesn't look too bad now that I actually did it.

mjordan commented on workbench:
```diff
@@ -672,9 +672,505 @@ def add_media():
+            else:
+                logging.warning("Field %s is not a valid plain-text field for media %s. This may be intended (e.g. media_id), but if not, please check the spelling of this field name.",
+                                field_name,
+                                media_type)
```
`media_id` and `media_use_tid` are being logged here as invalid plain text field names. While true, I'd like to see those two fields not be logged, since it's a bit misleading to the user (since they are reserved column names) and also adds extraneous entries to the log file. Can we skip logging the warning for these two column names?

hassanelsheikha replied:
Absolutely. I have committed the change now.

mjordan replied:
Thanks! Smoke tested and works great!

mjordan commented:
@hassanelsheikha your latest commit fixes the error about the missing `file` column. Thanks!

I haven't tested replacing media track files yet since it will take me a while to set that up, but I can offer a couple other pieces of feedback before I test that. This is awesome - thank you so much!

1. To be consistent with the `update` task on nodes, I'd like to see a message printed to the user and logged that says each media was updated or not, e.g. `Media https://localhost/media/168 updated` or `Media https://localhost/media/168 not updated. See log for more information.`
2. I think we should add a `ping_media` method to workbench_utils.py. Currently, `--check` passes but if the media does not actually exist, or if the `standalone_media_url` setting needs to be added to the config, the `update_media` task fails on the media. `--check` should ping each media so users know that there's a problem. See the `ping_node()` method for how Workbench does this for nodes.
3. Being able to patch plain text fields is an excellent start, but we will eventually need to support all field types (that Workbench supports). I can add this later, it's a bit of work. But there is another quirk to "updating" that we will want to add now - the "update mode" where the user defines whether the task "append", "replace", or "delete" the values. This is also quite a bit of work but the workbench_fields.py classes take care of most of this for you. I am happy to add that support later.

If you can take care of the first and second points above (messages to user and pinging media), I'll move on to testing replacing track files. Thanks!

mjordan commented:
I've added a comment over at https://github.com/mjordan/islandora_workbench_docs/pull/6/files to say we should add a warning to indicate that updating media replaces existing field values.

hassanelsheikha commented:
I will update the code ASAP (likely end of day today or tomorrow morning) for points 1 and 2. However, for point 3, would you say that this will be implemented in another PR or this one?

mjordan commented:
In a nutshell, for item 3 we will need to add logic [like this](https://github.com/mjordan/islandora_workbench/blob/main/workbench#L459-L487) where you call your new [patch_plain_text_fields()](https://github.com/mjordan/islandora_workbench/pull/602/files#diff-d85b40105dd0b9cdbe46a1ee1b96abdf1474ae96fb86bd28ca701f44f017ac3bR1133). The logic to handle the different `update_mode` values (replace, append, delete) is handled within the fields' `update` method. 

If you have time to take this on in this PR, please do! If you don't, no problem, I'm happy to implement item 3 since it might not be easy to follow what's going on.

hassanelsheikha commented:
I appreciate the opportunity. I don't think I'll be able to take on this task at the moment due to my current workload, but I will definitely get 1 and 2 done soon. I'm really sorry about any inconvenience this may cause you.

hassanelsheikha commented:
Thank you so much for your understanding. If you don't mind, I had a question about point 2 please. I actually created a `ping_media` function in the utils file:

https://github.com/hassanelsheikha/islandora_workbench/blob/5aa4169d2999b9a4f3dd4ee785d5561c9c75e17d/workbench_utils.py#L602

This function is used in the `update_media` routine to make sure that the referenced media object on the current row exists. If it doesn't, it will fail for that line and proceed to the next on (though I don't think it crashes; please correct me if I'm wrong). 

Is this the `ping_media` function you are expecting? If so, would you like me to modify the `--check` routine to make sure that all media objects referenced in the CSV are valid prior to the user running the actual `update_media` task? I would be happy to do this if you give the green light!

mjordan commented:
Adding/replacing media track file works as advertised.

mjordan commented:
@hassanelsheikha yes, sorry, I didn't see your ping_media function. That's great. Yes please, add to the check_input() function, just after line 1630, a new `if config['task'] == 'update_media':`  block and cycle through all the media_ids in the input CSV to ping them, and log if they don't exist.

mjordan commented:
Sorry, I closed by mistake, reopening.

hassanelsheikha commented:
Sorry for being a bit late on these commits. I updated the code to conform to your suggestions mentioned in points 1 and 2. The `--check` routine now scans every row to see if the media ID specified is valid. Also, the messages displayed to the user have been updated to say `Media at <HOSTNAME>/media/<MEDIA_ID> updated successfully` or `Media at  <HOSTNAME>/media/<MEDIA_ID> could not be updated. Please check the log for more details.`. There is an exception, however, which occurs when the user has specified an invalid media ID in the CSV. In this case, we can't print out the media ID as it is invalid or it is impossible to extract (in the case of a URL alias). So, I chose to print out `There are errors for CSV row <ROW NUMBER>. Please check the log for more details.` instead. I am happy to change this if it is not what you expected. Finally, I added a few more error handling pieces and updated the way `get_media_type` works (I didn't run into any bugs after this fix, but if future bugs arise I will fix it ASAP). 

If you have any feedback or other parts of the code you want changed, please let me know and I'll get to it right away. Thank you!
